File: leetcode/381-insert-delete-getrandom-o1-duplicates-allowed/main.go
package main

import (
	"container/list"
	"fmt"
	"math/rand"
)

/*
	learned from others
	- hashtable + linkedlist
	- https://www.youtube.com/watch?v=t5SwbbAfh-0
*/

type RandomizedCollection struct {
	rsm  map[int]*list.List
	keys []int
}

/** Initialize your data structure here. */
func Constructor() RandomizedCollection {
	return RandomizedCollection{rsm: make(map[int]*list.List), keys: make([]int, 0)}
}

/** Inserts a value to the collection. Returns true if the collection did not already contain the specified element. */
func (this *RandomizedCollection) Insert(val int) bool {
	var isPresent bool
	if _, ok := this.rsm[val]; !ok {
		isPresent = true
		this.rsm[val] = list.New()
	}
	this.rsm[val].PushBack(len(this.keys))
	this.keys = append(this.keys, val)

	return isPresent
}

/** Removes a value from the collection. Returns true if the collection contained the specified element. */
func (this *RandomizedCollection) Remove(val int) bool {
	if _, ok := this.rsm[val]; !ok {
		return false
	}

	// get the first value and remove it.
	valSet := this.rsm[val]
	loc := valSet.Front().Value.(int)
	valSet.Remove(valSet.Front())
	this.rsm[val] = valSet

	if loc < len(this.keys)-1 {
		lastOne := this.keys[len(this.keys)-1]
		this.keys[loc] = lastOne

		lastOneList := this.rsm[lastOne]
		lastOneList.Remove(lastOneList.Back())
		lastOneList.PushFront(loc)
		this.rsm[lastOne] = lastOneList
	}
	this.keys = this.keys[0 : len(this.keys)-1]

	if this.rsm[val].Len() == 0 {
		delete(this.rsm, val)
	}

	return true
}

/** Get a random element from the collection. */
func (this *RandomizedCollection) GetRandom() int {
	if len(this.keys) == 0 {
		return -1
	}
	return this.keys[rand.Intn(len(this.keys))]
}

func main() {
	c := Constructor()
	c.Insert(1)
	c.Insert(1)
	c.Insert(1)
	c.Insert(2)
	c.Insert(1)
	c.Insert(2)
	c.Insert(2)
	fmt.Println(c.rsm)
	fmt.Println(c.keys)
	c.Remove(1)
	fmt.Println(c.rsm)
	fmt.Println(c.keys)
	c.Remove(1)
	// c.Remove(1)
	// c.Remove(1)
	// c.Remove(1)
	fmt.Println(c.rsm)
	fmt.Println(c.keys)
	// c.Remove(1)
	// c.Remove(3)
}

File: leetcode/381-insert-delete-getrandom-o1-duplicates-allowed/main.py
import random


class RandomizedCollection(object):

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.m = {}

    def insert(self, val):
        """
        Inserts a value to the collection. Returns true if the collection did not already contain the specified element.
        :type val: int
        :rtype: bool
        """
        if val in self.m:
            self.m[val] += 1
            return False
        self.m[val] = 1
        return True

    def remove(self, val):
        """
        Removes a value from the collection. Returns true if the collection contained the specified element.
        :type val: int
        :rtype: bool
        """
        if val in self.m:
            self.m[val] -= 1
            if self.m[val] == 0:
                del self.m[val]
            return True
        return False

    def getRandom(self):
        """
        Get a random element from the collection.
        :rtype: int
        """
        arr = []
        for key in self.m:
            temp = [key]*self.m[key]
            arr += temp
        r = random.randint(0, len(arr)-1)
        return arr[r]


# Your RandomizedCollection object will be instantiated and called as such:
obj = RandomizedCollection()
print(obj.insert(1))
print(obj.insert(1))
print(obj.insert(2))
print(obj.getRandom())
print(obj.remove(1))
print(obj.remove(1))
print(obj.remove(1))
print(obj.remove(1))
print(obj.m)
print(obj.getRandom())
